crates/ruff/src/linter.rs

```diff
@@ -361,19 +361,26 @@ pub fn lint_only(
 
     // Convert from diagnostics to messages.
     let path_lossy = path.to_string_lossy();
+
     result.map(|(messages, imports)| {
+        let source_code = if settings.show_source && !messages.is_empty() {
+            Some(locator.to_source_code_buf())
+        } else {
+            None
+        };
+
         (
             messages
                 .into_iter()
                 .map(|diagnostic| {
-                    let source = if settings.show_source {
-                        Some(Source::from_diagnostic(&diagnostic, &locator))
-                    } else {
-                        None
-                    };
                     let lineno = diagnostic.location.row();
                     let noqa_row = *directives.noqa_line_for.get(&lineno).unwrap_or(&lineno);
-                    Message::from_diagnostic(diagnostic, path_lossy.to_string(), source, noqa_row)
+                    Message::from_diagnostic(
+                        diagnostic,
+                        path_lossy.to_string(),
```

Comment:
I wonder if we can apply some similar efficiency tricks to the path...? This also gets cloned and serialized once per diagnostic.

Reply:
Good idea. See #3904 (I created a PR on top to minimize the time spent rebasing the stacked PRs).